hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/OzoneProtocolMessageDispatcher.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ *  with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package org.apache.hadoop.hdds.server;
+
+import org.apache.hadoop.hdds.function.FunctionWithServiceException;
+import org.apache.hadoop.hdds.tracing.TracingUtil;
+import org.apache.hadoop.ozone.protocolPB.ProtocolMessageMetrics;
+
+import com.google.protobuf.ProtocolMessageEnum;
+import com.google.protobuf.ServiceException;
+import io.opentracing.Scope;
+import org.slf4j.Logger;
+
+/**
+ * Dispatch message after tracing and message logging for insight.
+ * <p>
+ * This is a generic utility to dispatch message in ServerSide translators.
+ * <p>
+ * It logs the message type/content on DEBUG/TRACING log for insight and create
+ * a new span based on the tracing information.
+ */
+public class OzoneProtocolMessageDispatcher<REQUEST, RESPONSE> {
+
+  private String serviceName;
+
+  private final ProtocolMessageMetrics protocolMessageMetrics;
+
+  private Logger logger;
+
+  public OzoneProtocolMessageDispatcher(String serviceName,
+      ProtocolMessageMetrics protocolMessageMetrics, Logger logger) {
+    this.serviceName = serviceName;
+    this.protocolMessageMetrics = protocolMessageMetrics;
+    this.logger = logger;
+  }
+
+  public RESPONSE processRequest(
+      REQUEST request,
+      FunctionWithServiceException<REQUEST, RESPONSE> methodCall,
+      ProtocolMessageEnum type,
+      String traceId) throws ServiceException {
+    Scope scope = TracingUtil
```

Comment:
Do we need to enclose this in try( Scope scope = TracingUtil.importAndCreateScope(type.toString(), traceId)) to release resources?

Reply:
There is a scope.close in the final branch.
This is the limitation of the auto-closable try catch block: with that approach the scope can't be used in the catch.

scope supports to mark the current scope as failed (which is not yet added to the code) which requires to follow this approach.

See [this](https://github.com/opentracing/opentracing-java) link:

> ScopeManager.active(Span, boolean) and SpanBuilder.startActive() have been deprecated as part of removing automatic Span finish upon Scope close, as doing it through try-with statements would make it hard to properly handle errors (Span objects would get finished before a catch block would be reached). This improves API safety, and makes it more difficult to do the wrong thing and end up with unexpected errors.